lib/api/v3/work_packages/schema/work_package_schemas_api.rb

```diff
@@ -56,25 +60,30 @@ def raise404
             namespace ':project-:type' do
               before do
                 begin
-                  project = Project.find(params[:project])
-                  type = Type.find(params[:type])
+                  @project = Project.find(params[:project])
+                  @type = Type.find(params[:type])
                 rescue ActiveRecord::RecordNotFound
                   raise404
                 end
 
-                authorize(:view_work_packages, context: project) do
+                authorize(:view_work_packages, context: @project) do
                   raise404
                 end
 
-                schema = TypedWorkPackageSchema.new(project: project, type: type)
-                self_link = api_v3_paths.work_package_schema(project.id, type.id)
-                @representer = WorkPackageSchemaRepresenter.create(schema,
-                                                                   self_link: self_link,
-                                                                   current_user: current_user)
+                # Compare with ETag composed of project and customizations
+                # to avoid evaluating the server request
+                @custom_fields = @project.all_work_package_custom_fields
+                with_etag! "#{@project.id}/#{@custom_fields.count}/#{@custom_fields.to_param}"
               end
 
               get do
-                @representer
+                cache([cache_key(@project.id, @type.id), @type.updated_at, @custom_fields]) do
+                  schema = TypedWorkPackageSchema.new(project: @project, type: @type)
+                  self_link = api_v3_paths.work_package_schema(@project.id, @type.id)
+                  WorkPackageSchemaRepresenter.create(schema,
+                                                      self_link: self_link,
+                                                      current_user: nil)
```

Comment:
Ah, good point. I would argue again that this is public information and does not need to be hidden behind a user check. All we know is that the project / work package allows to set costs. If the user is allowed to access the schema, that differentiation should not matter.

Thus I've introduced the `nil` here to make it clear that the schema representer may receive nil users for when showing the schema without its embedded form.

What do you think?